object_store_factory/src/aws.rs

```diff
@@ -0,0 +1,225 @@
+    aws::resolve_bucket_region, aws::AmazonS3Builder, aws::AmazonS3ConfigKey,
+    ClientOptions, ObjectStore,
+};
+use std::collections::HashMap;
+use std::env;
+use tracing::info;
+use url::Url;
+
+pub struct Config {
+    pub region: Option<String>,
+    pub access_key_id: Option<String>,
+    pub secret_access_key: Option<String>,
+    pub session_token: Option<String>,
+    pub endpoint: Option<String>,
+    pub bucket: String,
+    pub _prefix: Option<String>,
+}
+
+impl Config {
+    pub fn from_hashmap(
+        map: &HashMap<String, String>,
+    ) -> Result<Self, object_store::Error> {
+        Ok(Self {
+            region: map.get("region").map(|s| s.to_string()),
+            access_key_id: map.get("access_key_id").map(|s| s.to_string()),
+            secret_access_key: map.get("secret_access_key").map(|s| s.to_string()),
+            session_token: map.get("session_token").map(|s| s.to_string()),
+            endpoint: map.get("endpoint").map(|s| s.to_string()),
+            bucket: map.get("bucket").unwrap().clone(),
+            _prefix: map.get("prefix").map(|s| s.to_string()),
+        })
+    }
+}
+
+pub fn build_amazon_s3_from_config(
+    config: &Config,
+) -> Result<Box<dyn ObjectStore>, object_store::Error> {
+    let mut builder = AmazonS3Builder::new()
+        .with_region(config.region.clone().unwrap_or_default())
+        .with_bucket_name(config.bucket.clone())
+        .with_allow_http(true);
+
+    if let Some(endpoint) = &config.endpoint {
+        builder = builder.with_endpoint(endpoint.clone());
+    }
+
+    if let (Some(access_key_id), Some(secret_access_key)) =
+        (&config.access_key_id, &config.secret_access_key)
+    {
+        builder = builder
+            .with_access_key_id(access_key_id.clone())
+            .with_secret_access_key(secret_access_key.clone());
+
+        if let Some(token) = &config.session_token {
+            builder = builder.with_token(token.clone())
+        }
+    } else {
+        builder = builder.with_skip_signature(true)
+    }
+
+    let store = builder.build()?;
+    Ok(Box::new(store))
+}
+
+pub async fn add_amazon_s3_specific_options(
+    url: &Url,
+    mut options: HashMap<String, String>,
+) -> HashMap<String, String> {
+    if !options.contains_key(AmazonS3ConfigKey::Bucket.as_ref())
+        && !options.contains_key(AmazonS3ConfigKey::Endpoint.as_ref())
+    {
+        let region = detect_region(url).await.unwrap();
+        options.insert("region".to_string(), region.to_string());
+    }
+    options
+}
+
+pub fn add_amazon_s3_environment_variables(
+    mut options: HashMap<String, String>,
+) -> HashMap<String, String> {
+    let env_vars = &[
+        ("AWS_ACCESS_KEY_ID", "access_key_id"),
+        ("AWS_SECRET_ACCESS_KEY", "secret_access_key"),
+        ("AWS_DEFAULT_REGION", "region"),
+        ("AWS_ENDPOINT", "endpoint"),
+        ("AWS_SESSION_TOKEN", "session_token"),
+        ("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI", "credentials_uri"),
```

Comment:
You could use `AmazonS3ConfigKey` for these as well (likewise for other stores).